File: majorProject/src/main/java/com/aocc/majorproject/GamePreferences.java
package com.aocc.majorproject;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.AudioManager;

import com.aocc.framework.implementation.RotationHandler;
import com.aocc.framework.PersonalMethods;

/**
 * Persisted player preferences. Loaded once at startup; saved on each change.
 */
public final class GamePreferences {

    private static final String PREFS_NAME = "major_project_prefs";
    private static final String KEY_SOUND = "sound_enabled";
    private static final String KEY_MUSIC = "music_enabled";
    private static final String KEY_MUSIC_INITIALIZED = "music_initialized";
    private static final String KEY_TILT_MODE = "tilt_mode";
    private static final String KEY_SECOND_SCREEN = "second_screen_enabled";

    private static SharedPreferences prefs;

    public static boolean sound = true;
    public static boolean music = true;
    public static int tiltMode = 2;
    public static boolean secondScreenEnabled = false;

    private GamePreferences() {
    }

    public static void load(Context context) {
        prefs = context.getApplicationContext()
                .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        sound = prefs.getBoolean(KEY_SOUND, true);
        tiltMode = prefs.getInt(KEY_TILT_MODE, 2);
        secondScreenEnabled = prefs.getBoolean(KEY_SECOND_SCREEN, false);

        if (!prefs.getBoolean(KEY_MUSIC_INITIALIZED, false)) {
            AudioManager audioManager =
                    (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
            music = audioManager == null || !audioManager.isMusicActive();
            prefs.edit()
                    .putBoolean(KEY_MUSIC, music)
                    .putBoolean(KEY_MUSIC_INITIALIZED, true)
                    .apply();
        } else {
            music = prefs.getBoolean(KEY_MUSIC, true);
        }
    }

    public static int getTapVolume() {
        return sound ? 10 : 0;
    }

    public static void setSound(boolean enabled) {
        sound = enabled;
        save();
    }

    public static void setMusic(boolean enabled) {
        music = enabled;
        save();
    }

    public static void setTiltMode(int mode) {
        tiltMode = mode;
        save();
    }

    public static void setSecondScreenEnabled(boolean enabled) {
        secondScreenEnabled = enabled;
        save();
    }

    public static void applyTiltTo(Player player) {
        switch (tiltMode) {
            case 1:
                player.setxBias(0);
                player.setyBias(0);
                player.setTiltMode(1);
                break;
            case 3:
                player.setxBias(-PersonalMethods.limitInside(RotationHandler.getRotationX(), -90, 90) / 90);
                player.setyBias(-PersonalMethods.limitInside(RotationHandler.getRotationY(), -90, 90) / 90);
                player.setTiltMode(3);
                break;
            case 2:
            default:
                player.setxBias(0);
                player.setyBias(-0.30f);
                player.setTiltMode(2);
                break;
        }
    }

    private static void save() {
        if (prefs == null) {
            return;
        }
        prefs.edit()
                .putBoolean(KEY_SOUND, sound)
                .putBoolean(KEY_MUSIC, music)
                .putInt(KEY_TILT_MODE, tiltMode)
                .putBoolean(KEY_SECOND_SCREEN, secondScreenEnabled)
                .apply();
    }
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/GamePreferences.kt
package com.aocc.majorproject

import android.content.Context
import android.media.AudioManager
import com.aocc.framework.PersonalMethods
import com.aocc.framework.implementation.RotationHandler

/**
 * Persisted player preferences. Loaded once at startup; saved on each change.
 */
object GamePreferences {

    private const val PREFS_NAME = "major_project_prefs"
    private const val KEY_SOUND = "sound_enabled"
    private const val KEY_MUSIC = "music_enabled"
    private const val KEY_MUSIC_INITIALIZED = "music_initialized"
    private const val KEY_TILT_MODE = "tilt_mode"
    private const val KEY_SECOND_SCREEN = "second_screen_enabled"

    private var prefs: android.content.SharedPreferences? = null

    @JvmField
    var sound: Boolean = true

    @JvmField
    var music: Boolean = true

    @JvmField
    var tiltMode: Int = 2

    @JvmField
    var secondScreenEnabled: Boolean = false

    @JvmStatic
    fun load(context: Context) {
        val loaded = context.applicationContext
            .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs = loaded
        sound = loaded.getBoolean(KEY_SOUND, true)
        tiltMode = loaded.getInt(KEY_TILT_MODE, 2)
        secondScreenEnabled = loaded.getBoolean(KEY_SECOND_SCREEN, false)

        if (!loaded.getBoolean(KEY_MUSIC_INITIALIZED, false)) {
            val audioManager =
                context.getSystemService(Context.AUDIO_SERVICE) as AudioManager?
            music = audioManager == null || !audioManager.isMusicActive
            loaded.edit()
                .putBoolean(KEY_MUSIC, music)
                .putBoolean(KEY_MUSIC_INITIALIZED, true)
                .apply()
        } else {
            music = loaded.getBoolean(KEY_MUSIC, true)
        }
    }

    @JvmStatic
    fun getTapVolume(): Int {
        return if (sound) 10 else 0
    }

    @JvmStatic
    fun setSound(enabled: Boolean) {
        sound = enabled
        save()
    }

    @JvmStatic
    fun setMusic(enabled: Boolean) {
        music = enabled
        save()
    }

    @JvmStatic
    fun setTiltMode(mode: Int) {
        tiltMode = mode
        save()
    }

    @JvmStatic
    fun setSecondScreenEnabled(enabled: Boolean) {
        secondScreenEnabled = enabled
        save()
    }

    @JvmStatic
    fun applyTiltTo(player: Player) {
        when (tiltMode) {
            1 -> {
                player.xBias = 0f
                player.yBias = 0f
                player.tiltMode = 1
            }
            3 -> {
                player.xBias =
                    -PersonalMethods.limitInside(RotationHandler.getRotationX(), -90, 90) / 90f
                player.yBias =
                    -PersonalMethods.limitInside(RotationHandler.getRotationY(), -90, 90) / 90f
                player.tiltMode = 3
            }
            else -> {
                player.xBias = 0f
                player.yBias = -0.30f
                player.tiltMode = 2
            }
        }
    }

    private fun save() {
        prefs?.edit()
            ?.putBoolean(KEY_SOUND, sound)
            ?.putBoolean(KEY_MUSIC, music)
            ?.putInt(KEY_TILT_MODE, tiltMode)
            ?.putBoolean(KEY_SECOND_SCREEN, secondScreenEnabled)
            ?.apply()
    }
}
